File: Python/Algorithms/133.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        # keep track of what we've already visited as
        # a mapping from node values to (cloned) nodes.
        visited = dict()

        def clone(node):
            # return the created node if it has already been
            # visited.
            if node.val in visited:
                return visited[node.val]
            # create node *first* w/o neighbors, add to cache
            # and *then* recurse.
            n = Node(node.val)
            visited[n.val] = n
            n.neighbors = [clone(n) for n in node.neighbors]
            return n
        return clone(node) if node else node

File: Javascript/Algorithms/133.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
  const visited = Array(101);
  const clone = node => {
    if (visited[node.val]) return visited[node.val];
    let neighbors = [];
    // pass ref to neighbors over to new node.
    let n = new Node(node.val, neighbors);
    visited[n.val] = n;
    node.neighbors.forEach(neighbor => neighbors.push(clone(neighbor)));
    return n;
  };
  return node ? clone(node) : null;
};